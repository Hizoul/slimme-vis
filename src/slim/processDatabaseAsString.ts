
export interface IClass {
  label: string
  values: number
}

export interface Database {[index: string]: number[][] | undefined}

const processDatabaseAsString = (databaseContent: string) => {
  const classes: Database = {}
  for (const line of databaseContent.split("\n")) {
    const values = line.split(" ").map((value) => {
      if (value.length > 0) {
        return Number(value)
      }
    }).filter((val) => val != null)
    const classKey = String(values[0])
    if (classes[classKey] == null) {
      classes[classKey] = []
    }
    classes[classKey].push(values.splice(1))
  }
  return classes
}

export default processDatabaseAsString
