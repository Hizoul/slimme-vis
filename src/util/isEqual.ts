const isEqual = (a: any, b: any) => {
  if (Array.isArray(a) && Array.isArray(b)) {
    if (a.length !== b.length) {
      return false
    }
    for (let i = 0; i < a.length; i++) {
      if (a[i] !== b[i]) {
        return false
      }
    }
    return true
  }
  return a === b
}

export default isEqual
