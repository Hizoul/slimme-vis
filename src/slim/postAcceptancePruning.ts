import { debug } from "util"
import isEqual from "../util/isEqual"
import isValidTable from "../util/isValidTable"
import { codeLengthIncludingStandard } from "./codeLength"
import { Database } from "./processDatabaseAsString"
import { SortedCodeTable, updateFrequencies } from "./standardCodeTable"

const findDecreasingEntries = (oldTable: SortedCodeTable, newTable: SortedCodeTable) => {
  const toReconsider: SortedCodeTable = []
  for (const entry of newTable) {
    SEARCH: for (const oldEntry of oldTable) {
      if (entry[0].length > 1 && isEqual(entry[0], oldEntry[0])) {
        if (entry[1][0] < oldEntry[1][0]) {
          toReconsider.push(entry)
        }
        break SEARCH
      }
    }
  }
  toReconsider.sort((a, b) => {
    // if (a.frequency === 0) {
    //   return 1
    // }
    // if (b.frequency === 0) {
    //   return -1
    // }
    return a[1][0] - b[1][0]
  })
  return toReconsider
}

const postAcceptancePruning =
(dataset: Database, oldTable: SortedCodeTable, newTable: SortedCodeTable, standardTable: SortedCodeTable) => {
  const toReconsider = findDecreasingEntries(oldTable, newTable)
  let tableToReturn: SortedCodeTable = newTable
  let previousTable: SortedCodeTable = newTable
  while (toReconsider.length > 0) {
    const ele = toReconsider.shift()
    let tmpTable = tableToReturn.filter((entry) => !isEqual(ele[0], entry[0]))
    tmpTable = updateFrequencies(dataset, tmpTable)
    if (codeLengthIncludingStandard(dataset, tmpTable, standardTable) <
      codeLengthIncludingStandard(dataset, tableToReturn, standardTable) && isValidTable(dataset, tmpTable)) {
        previousTable = tableToReturn
        tableToReturn = tmpTable
        const newlyConsidred = findDecreasingEntries(previousTable, tableToReturn)
        for (const item of newlyConsidred) {
          if (toReconsider.filter((e) => isEqual(e[0], item[0])).length === 0) {
            toReconsider.push(item)
          }
        }
    }
  }
  tableToReturn = tableToReturn.filter((entry) => {
    return entry[1][0] > 0
  })
  return tableToReturn
}

export default postAcceptancePruning
