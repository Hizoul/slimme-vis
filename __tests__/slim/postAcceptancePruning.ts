import postAcceptancePruning from "../../src/slim/postAcceptancePruning"
import processDatabaseAsString from "../../src/slim/processDatabaseAsString"
import sortedStandardCodeTable, { SortedCodeTable } from "../../src/slim/standardCodeTable"
import isEqual from "../../src/util/isEqual"
const database = `
0 1 2 3
0 1 2 3
0 1 2 3
0 1 2 3
0 1 2 3
0 1 2
0 1
0 2
`

const codeTable: SortedCodeTable = [
  [[1, 2, 3], [5]],
  [[1, 2], [1]],
  [[1], [1]],
  [[2], [1]],
  [[3], [0]]
]
const previousTable: SortedCodeTable = [
  [[1, 2], [6]],
  [[3], [5]],
  [[1], [1]],
  [[2], [1]]
]

describe("Post Acceptance Pruning", () => {
  it("should reduce codelength", () => {
    const db = processDatabaseAsString(database)
    const standardTable = sortedStandardCodeTable(db)
    const prunedTable = postAcceptancePruning(db, previousTable, codeTable, standardTable)
    expect(isEqual(prunedTable, codeTable)).toBe(false)
    expect(isEqual(prunedTable, previousTable)).toBe(false)
    expect(prunedTable).toMatchSnapshot("pruned table")
  })
})
