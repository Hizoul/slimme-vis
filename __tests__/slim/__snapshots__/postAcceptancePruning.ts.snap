// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post Acceptance Pruning should reduce codelength: pruned table 1`] = `
Array [
  Array [
    Array [
      1,
      2,
      3,
    ],
    Array [
      5,
    ],
  ],
  Array [
    Array [
      2,
    ],
    Array [
      2,
    ],
  ],
  Array [
    Array [
      1,
    ],
    Array [
      2,
    ],
  ],
]
`;
